### PR TITLE
chore(protocol): docs how to charge ERC20 token as proving fee

### PR DESCRIPTION
**Pull request by adaki2004**
High level description how one can create a prover pool which accepts any arbitrary tokens.

### CODE REVIEW

File: packages/protocol/docs/iprover.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+The proposer and prover interact off-chain to agree on the price and perform the ERC-20 token approval. Here's an example of how this interaction might work:
+
+1. Proposer asks the ProverPool for the cost of proving a block and receives a price (e.g., `10 DAI` tokens). If price is acceptable, prover provides proposer with a valid ECDSA signature which signs the commitment (e.g.: transaction list hash), expiration and price.
+
+2. Proposer signs the following transaction off-chain: `ERC-20(DAI_ADDRESS).approve(proverPool, 10)`.
+
+3. Proposer encodes these signatures into the `assignment.data` and calls the `proposeBlock()`.
+
+During `proposeBlock()` transaction, the `onBlockAssigned()` hook which will evaluate the validity of the prover signature, and if that one is correct then executes 2 transactions:
+1. The approval `ERC-20(DAI_ADDRESS).approve(proverPool, 10)` which is signed by the proposer.
```
Comment (dantaik) [inline packages/protocol/docs/iprover.md:64]: I think the `approve` transaction should be independent of the `proposeBlock` transaction and should happen before-ward.

File: packages/protocol/docs/iprover.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        TaikoData.BlockMetadataInput calldata input,
+        TaikoData.ProverAssignment calldata assignment
+    ) external {
+        // Decode the assignment data to retrieve signatures and other information
+        (bytes memory proverSignature, bytes memory signedApproveTxn, uint256 tokenAmount) = decodeAssignmentData(assignment);
+
+        // 1. Verify the prover signature is valid (off-chain verification)
+        require(isValidSignature(proverSignature, input), "Invalid prover signature");
+
+        // 2. Execute approve
```
Comment (dantaik) [inline packages/protocol/docs/iprover.md:40]: You cannot approve tokens inside a contract, msg.sender is not the owner.

**Review by dantaik — DISMISSED** [review]

**Review by dantaik — DISMISSED** [review]

File: packages/protocol/docs/iprover.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        TaikoData.ProverAssignment calldata assignment
+    ) external {
+        // Decode the assignment data to retrieve signatures and other information
+        (bytes memory proverSignature, uint256 tokenAmount) = decodeAssignmentData(assignment);
+
+        // 1. Verify the prover signature is valid (off-chain verification)
+        require(isValidSignature(proverSignature, input), "Invalid prover signature");
+
+        // 2. Execute the transfer transaction
+        ERC20(ERC20TokenAddress).transferFrom(input.proposer, address(this), tokenAmount);
```
Comment (davidtaikocha) [inline packages/protocol/docs/iprover.md:41]:
```suggestion
        ERC20(ERC20TokenAddress).transferFrom(msg.sender, address(this), tokenAmount);
```
would suggest `msg.sender` to give out the ERC-20 tokens, `input.proposer` is the L2 account which will be corresponding L2 blocks's coinbase.

Comment (adaki2004) [reply packages/protocol/docs/iprover.md:41]: Ok, i'd then add `tx.origin` bc. `msg.sender` is the TaikoL1.

**Review by davidtaikocha — APPROVED** [review]

**Review by dantaik — APPROVED** [review]

**Pull request merged**